### PR TITLE
Support work stealing for resource restricted tasks (see issue #1389)

### PR DESCRIPTION
**Pull request by quakberto**
(Addresses issue #1389)

Idea is to naturally extend the current stealing approach: If a key has resource restrictions, choose the next idle worker providing the required resources rather than simply the next idle worker.

Modifications in stealing.py:
- find next worker providing the necessary resources
- make sure resources are properly consumed and released during stealing

Added tests to verify that
- resource restricted tasks are stolen when workers with sufficient resources are available (and resources are properly consumed and released)
- resource restricted tasks are not stolen when available workers do not possess sufficient resources

Mentioned the support of the resource framework in the documentation.

### CODE REVIEW

File: distributed/stealing.py [hunk 1]
```diff
@@ -331,6 +346,19 @@ def balance(self):
...
+                        # set i to next worker providing them
+                        if key in self.scheduler.resource_restrictions.keys():
+                            resource_i = self.next_resource_i(
+                                idle,
+                                self.scheduler.resource_restrictions[key],
+                                start_index=i)
+                            if resource_i is None:
+                                continue
+                            else:
+                                i = resource_i
```
Comment (mrocklin) [inline distributed/stealing.py]: Maybe this logic should just return the worker to steal from, rather than return an index into the idle collection.

**Comment by quakberto** [comment]
Thanks for your comments!

I understand that you are concerned about performance. I guess as you said the danger would be to have tasks that are difficult to steal (for these tasks the for loop searching for a suitable worker could iterate through the entire idle collection each time).

Maybe one way to optimise this would be to keep track of available resources provided by the idle workers? Searching for the thief could be implemented as a binary tree search. The additionally required computation time would then scale better with the amount of resource restricted workers.

Could you explain what you mean by ensuring the balancing of load?

File: distributed/stealing.py [hunk 2]
```diff
@@ -234,6 +234,8 @@ def move_task_confirm(self, key=None, worker=None, state=None):
 
                 try:
                     self.scheduler.send_task_to_worker(thief, key)
+                    self.scheduler.consume_resources(key, thief)
+                    self.scheduler.release_resources(key, victim)
```
[the remark this answers is not in the document]
Comment (quakberto) [reply distributed/stealing.py]: True, if sending to the thief fails the victim should still release the resources and the thief is removed anyways.

File: distributed/stealing.py [hunk 3]
```diff
@@ -288,6 +290,19 @@ def balance(self):
                             stealable.remove(key)
                             continue
                         i += 1
+
+                        # in case the key has resource_restrictions,
+                        # set i to next worker providing them
+                        if key in self.scheduler.resource_restrictions.keys():
```
[the remark this answers is not in the document]
Comment (quakberto) [reply distributed/stealing.py]: Clear.

File: distributed/stealing.py (hunk 1 above)
Comment (quakberto) [reply distributed/stealing.py]: I tried to be minimally intrusive and stick to increasing the starting point for worker search...

**Comment by quakberto** [comment]
I've been trying to understand why the test `test_submit_many_non_overlapping` in `test_resources.py` is failing for my new branch (warning: there are two test functions with the same name and one overrides the other one). The critical assert is one verifying that after waiting for all futures the total resources of each worker equal the available ones.

It seems like the test sometimes passes and sometimes not. Comparing the logs between successful and unsuccessful runs shows that in the unsuccessful case, a key is submitted for execution after the cluster is closed, as shown in this log:

```
...
distributed.scheduler - DEBUG - Finished handle_client coroutine
distributed.worker - DEBUG - Send compute response to scheduler: slowinc-5351d71bc9c5fa2f3a8e4aaadb1b8e49, {'status': 'OK', 'thread': 123145342308352, 'stop': 1510137945.6390312, 'start': 1510137945.615948, 'nbytes': 24, 'key': 'slowinc-5351d71bc9c5fa2f3a8e4aaadb1b8e49', 'op': 'task-finished', 'type': <type 'int'>}
distributed.utils_test - DEBUG - Closing out test cluster
distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:56481
distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:56483
distributed.batched - INFO - Batched Comm Closed:
distributed.worker - DEBUG - Execute key: slowinc-fe44027e7d70370c034ef17cb082e33d worker: tcp://127.0.0.1:56483
distributed.worker - ERROR - Thread Pool Executor error: cannot schedule new futures after shutdown
distributed.worker - DEBUG - Deleted 1 keys
...
```

Counting how often `transition_constrained_executing` and `transition_executing_done` are called in this case shows that `transition_executing_done` is lagging behind by one. I guess this leads to `self.available_resources[resource] += quantity` not being executed (meaning unreleased resources).

Unfortunately, evaluating this late task execution goes beyond my current understanding of distributed...

**Comment by mrocklin** [comment]
This PR has gone stale.  Closing for now.

**Pull request closed without merging**